src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs

```diff
@@ -2316,7 +2317,7 @@ private BoundExpression BindUnaryOperatorCore(CSharpSyntaxNode node, string oper
             }
 
             // If the operand is bad, avoid generating cascading errors.
-            if (operand.HasAnyErrors || isOperandTypeNull)
+            if (isOperandTypeNull || operand.Type?.IsErrorType() == true)
```

Comment:
The root of the bug is that, when speculating on a pattern that introduces a variable that conflicts with an existing variable, the entire declaration pattern is marked as errored as it introduced a name conflict. We then bail early here, creating a `BoundUnaryOperator` with type `ErrorType`. This change narrows the scope of where we bail out to only bail if the operand bound to an error type. This could produce a few more cascading errors, but I didn't see any additional failures in the compiler tests and if we run into more examples in the real world, we can narrow further at that point.